packages/grafana-ui/src/types/datasource.ts

```diff
@@ -164,21 +168,28 @@ export interface DataSourceApi<
    * Set after constructor call, as the data source instance is the most common thing to pass around
    * we attach the components to this instance for easy access
    */
-  components?: DataSourcePluginComponents<TQuery, TOptions>;
+  components?: DataSourcePluginComponents<DataSourceApi<TQuery, TOptions>, TQuery, TOptions>;
 
   /**
    * static information about the datasource
    */
   meta?: DataSourcePluginMeta;
 }
 
-export interface ExploreDataSourceApi<TQuery extends DataQuery = DataQuery> extends DataSourceApi {
+export interface ExploreDataSourceApi<
+  TQuery extends DataQuery = DataQuery,
+  TOptions extends DataSourceJsonData = DataSourceJsonData
+> extends DataSourceApi<TQuery, TOptions> {
   modifyQuery?(query: TQuery, action: QueryFixAction): TQuery;
   getHighlighterExpression?(query: TQuery): string;
   languageProvider?: any;
 }
 
-export interface QueryEditorProps<DSType extends DataSourceApi, TQuery extends DataQuery> {
+export interface QueryEditorProps<
+  DSType extends DataSourceApi<TQuery, TOptions>,
```

Comment:
It is subtle, but this is the key change -- everything else just makes this work:

The DSType before was `DSType extends DataSourceApi` -- but we need it to be `DSType extends DataSourceApi<TQuery, TOptions>`